internal/guest/transport/devnull.go

```diff
@@ -0,0 +1,132 @@
+//go:build linux
+// +build linux
+
+package transport
+
+import (
+	"os"
+
+	"github.com/sirupsen/logrus"
+)
+
+// DevNullTransport is a transport that will:
+//
+// For reads: return either closed or EOF for as appropriate.
+// For writers: return either closed or throw away the write as appropriate.
+//
+// The DevNullTransport is used as the container logging transport when stdio
+// access is denied. It is also used for non-terminal external processes (aka
+// a process run in the UVM) when stdio access is denied.
+type DevNullTransport struct{}
+
+func (t *DevNullTransport) Dial(fd uint32) (Connection, error) {
+	logrus.WithFields(logrus.Fields{
+		"fd": fd,
+	}).Info("opengcs::DevNullTransport::Dial")
+
+	return newDevNullConnection(), nil
+}
+
+// devNullConnection is the heart of our new transport. A devNullConnection
+// contains two file descriptors. One for read, and one for write. We need to
+// file descriptors as the code using a transport is written with the
+// expectation of a duplex connection where read and write can be closed
+// independent of one another. The protocol that uses the connection requires
+// a duplex connection in order to work correctly.
+//
+// The original design of devNullConnection didn't use os.File and instead
+// emulated the required behavior and only returned a os.File from the `File`
+// function. However, the amount of state required to manage that was somewhat
+// high and the logic to follow was somewhat convoluted. Using two file handles
+// one read, one write that have /dev/null open results in a much cleaner
+// design and easier to understand semantics. We want to mirror the sematics of
+// /dev/null as a duplex connection, this is the simplest, most understandable
+// way to achieve that.
+type devNullConnection struct {
+	read  *os.File
+	write *os.File
+}
+
+func newDevNullConnection() *devNullConnection {
+	r, _ := os.OpenFile(os.DevNull, os.O_RDONLY, 0)
+	w, _ := os.OpenFile(os.DevNull, os.O_WRONLY, 0)
+
+	return &devNullConnection{read: r, write: w}
```

Comment:
It would be one fewer file handle to open this with a single `os.OpenFile(...)` call and use io.Reader / io.Writer in the devNullConnection both pointing to the same file.  CloseRead and Close write could just set the corresponding item to null.  Then the File() at the bottom could also just return this same os.File as is opened in new here.